payments/schema/mutations.py

```diff
@@ -741,6 +747,118 @@ def mutate_and_get_payload(cls, root, info, order_id, **input):
         return RefundOrderMutation(order_refund=refund)
 
 
+class CreateBerthSwitchOfferMutation(graphene.ClientIDMutation):
+    class Input:
+        # Required fields
+        application_id = graphene.ID(required=True)
+        new_berth_id = graphene.ID(required=True)
+        # Optional fields
+        old_lease_id = graphene.ID(
+            description="If provided, it will ignore the `Berth Switch` from the application "
+            "and force this lease to be terminated"
+        )
+        due_date = graphene.Date()
+        profile_token = graphene.String(
+            description="API token for Helsinki profile GraphQL API",
+        )
+
+    berth_switch_offer = graphene.Field(BerthSwitchOfferNode, required=True)
+
+    @staticmethod
+    def get_old_lease(application: BerthApplication) -> BerthLease:
+        # TODO: This logic has to be refactored when the old harbors app is removed
+
+        # Based on the information filled by the customer on the switch application,
+        # we retrieve the corresponding lease
+        switch: BerthSwitch = application.berth_switch
+        berth_filters = Q(number__iexact=switch.berth_number)
+
+        # First check if the chosen harbor has a direct map with a resources harbor
+        if harbor := switch.harbor.resources_harbor:
+            berth_filters &= Q(pier__harbor=harbor)
+
+            # Try to match the one with the identifier provided by the customer
+            pier = harbor.piers.filter(identifier__iexact=switch.pier).first()
+
+            # If no pier is found but the harbor only has one pier associated,
+            # we can assume that the old berth has to be on that one pier
+            if not pier and harbor.piers.count() == 1:
+                pier = harbor.piers.first()
+
+            berth_filters &= Q(pier=pier)
+        else:
+            # Fetch all the piers associated to the selected harbor,
+            # in most cases there will be only one pier
+            piers = switch.harbor.resources_pier.all()
+
+            # Try to match the one with the identifier provided by the customer
+            pier = piers.filter(identifier__iexact=switch.pier).first()
+
+            # If no pier is found but the harbor only has one pier associated,
+            # we can assume that the old berth has to be on that one pier
+            if not pier and piers.count() == 1:
+                pier = piers.first()
+
+            berth_filters &= Q(pier=pier)
+
+        old_berth = Berth.objects.get(berth_filters)
+
+        # Find the lease on the current season
+        return BerthLease.objects.get(
+            customer=application.customer,
+            berth=old_berth,
+            status=LeaseStatus.PAID,
+            start_date=calculate_season_start_date(),
+            end_date=calculate_season_end_date(),
```

Comment:
Thinking of a special case. Are berth switch offers ever created in the time range from September 14th to December 31th?
Let's say, for some reason, that the admin wants to create a switch offer sometime in December 2021.
In that case, the 2021 season has already ended and in order to proceed with the switch, the customer needs to have a paid lease for the 2022 season. but calculate_season_(start|end)_date would return dates from 2021. So we'd need here functions like "calculate_current_or_upcoming_season_(start|end)_date"?

Actually I'm not entirely sure if this case would ever occur and if I understand the process correctly, but maybe the rules should be:
* If switch offer is created while season is active, the user gets the new berth after accepting the offer. Customer must have paid lease for the ongoing season.
* If switch offer is created at any other time, then the user gets the new berth when the next season starts. Customer must have paid lease for the next season.

Reply:
Joonas and Kalle commented in slack that it's very unlikely that switch offers would be created in days between season's end and the start of the next year, so this question is kind of moot. Maybe just prevent creating the switch offer if the lease.end_date < today()?

Reply:
> but calculate_season_(start|end)date would return dates from 2021. So we'd need here functions like "calculate_current_or_upcoming_season(start|end)_date"?

The calculate season start/end check if the season has already ended (e.g. between September and December). If so, they return the dates for the next season.
```python
if today >= date(day=14, month=9, year=today.year):
    return default.replace(year=today.year + 1)
```

Reply:
OK! I got confused when I tried:
`
In [5]: calculate_season_start_date(datetime.date(2020,11,1))
Out[5]: datetime.date(2020, 6, 10)
`
But now I see that calling the function without argument indeed fixes it.